File: docs/JAMIA/comparisons/RPython/TM/29224719_ifelse_na_type_conflict.ts
/**
 * RPython SO#29224719 — ifelse with NA causes logical vs numeric type conflict
 * Effect: Crash
 * Bug class: Nullable type
 *
 * R bug: ifelse(hits.diff <= 0, -hits.diff, 0) returns logical NA when all values
 * in a group are NA (because NA <= 0 is NA, so ifelse returns its "no" branch type
 * for the first two rows which are all NA → logical). When group_by produces a group
 * where some rows are numeric and the next group is logical, dplyr crashes:
 * "incompatible types, expecting a logical vector."
 *
 * In tidy-ts, a summarize/mutate that may return null produces a nullable type.
 * Using that nullable result in Math.round (which requires strict number) is caught.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const df1 = createDataFrame([
  { crawlId: 1, groupId: "1", hitsDiff: null as number | null },
  { crawlId: 1, groupId: "2", hitsDiff: null as number | null },
  { crawlId: 2, groupId: "2", hitsDiff: 0 },
  { crawlId: 1, groupId: "3", hitsDiff: null as number | null },
  { crawlId: 1, groupId: "3", hitsDiff: null as number | null },
  { crawlId: 1, groupId: "3", hitsDiff: null as number | null },
]);

// The SO user's intent: compute hits.consumed = ifelse(hits.diff <= 0, -hits.diff, 0)
// per group. R crashes because NA propagates and changes the return type to logical.
// tidy-ts: the column is number | null. Mutating with a function that doesn't handle
// null produces number | null. Using that in Math.round is rejected.
const result = df1.mutate({
  hitsConsumed: (r) => (r.hitsDiff !== null && r.hitsDiff <= 0) ? -r.hitsDiff : null,
});

// The result column is number | null. Passing to Math.round (requires number) fails.
// @ts-expect-error — Argument of type 'number | null' is not assignable to parameter of type 'number'
result.mutate({ rounded: (r) => Math.round(r.hitsConsumed) });

File: docs/JAMIA/comparisons/RPython/TM/29224719_ifelse_na_type_conflict.R
# ID: SO#29224719
# Language: R
# Bug class: Nullable
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: All-NA group makes ifelse return logical instead of numeric; type conflict across groups
# Reproduction status: Fixed (dplyr 1.1+ handles NA type coercion)
# Type system catch: `number` rejects `number | null`

library(dplyr)

df1 <- data.frame(
  crawl.id = c(1, 1, 2, 1, 1, 1),
  group.id = factor(c("1", "2", "2", "3", "3", "3")),
  hits.diff = c(NA, NA, 0, NA, NA, NA)
)

cat("Input data:\n")
print(df1)
cat("\n")

# The bug: when all hits.diff in a group are NA, ifelse returns logical NA
# instead of numeric 0. dplyr crashes on type mismatch between groups.
tryCatch({
  result <- df1 %>%
    group_by(group.id) %>%
    mutate(hits.consumed = ifelse(hits.diff <= 0, -hits.diff, 0))
  print(result)
}, error = function(e) {
  cat(paste("Crash:", e$message, "\n"))
})
